### PR TITLE
Add ephesoft_connector

### PR DESCRIPTION
**Pull request by Ephesoft-Stitus**
# What
Adding the Ephesoft Semantik For Invoices connector.

# How Was This Tested
Manually deployed as a custom connector and tested manually.

### CODE REVIEW

File: custom_connectors/oauth2/ephesoft_connector.rb [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+                        client_secret: connection["client_secret"],
+                        redirect_uri: "https://www.workato.com/oauth/callback"
+                      ).
+                      request_format_www_form_urlencoded
+        [
+          { # Tokens hash
+            access_token: response["access_token"],
+            refresh_token: response["refresh_token"]
+          },
+          { instance_id: nil } # Optional. Will be merged into connection hash
```
Comment (bennettgo) [inline custom_connectors/oauth2/ephesoft_connector.rb:64]: This isn't necessary. We could probably remove it! Only needed if you want to include additional attributes.

File: custom_connectors/oauth2/ephesoft_connector.rb [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    lambda do |connection|
+      get("https://api.us.ephesoft.io/v1/settings/integrations/configurations")
+    end,
+  actions: {
+    # Future action code here
+  },
+  triggers: {
+    new_message: {
+      title: 'When a Semantik invoice has completed review',
+      description: 'Triggers a flow using data from a completed Semantik invoice.',
```
Comment (bennettgo) [inline custom_connectors/oauth2/ephesoft_connector.rb:83]: Normally we follow this terminology:
`Triggers when a Semantik Invoice is completed`

File: custom_connectors/oauth2/ephesoft_connector.rb [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+  test:
+    lambda do |connection|
+      get("https://api.us.ephesoft.io/v1/settings/integrations/configurations")
+    end,
+  actions: {
+    # Future action code here
+  },
+  triggers: {
+    new_message: {
+      title: 'When a Semantik invoice has completed review',
```
Comment (bennettgo) [inline custom_connectors/oauth2/ephesoft_connector.rb:82]: Titles are normally in this format:
`New complete Semantik Invoice`

File: custom_connectors/oauth2/ephesoft_connector.rb [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
...\n     \"TaxAmount\": \"$TaxAmount\",\n     \"TaxId\": \"$TaxId\",\n     \"TaxRate\": \"$TaxRate\",\n     \"TenantId\": \"$TenantId\",\n     \"Terms\": \"$Terms\",\n     \"TotalAmount\": \"$TotalAmount\",\n     \"Vendor\": {\n       \"VendorID\": \"$Vendor:VendorID\",\n       \"VendorName\": \"$Vendor:Name\",\n       \"VendorAddress\": {\n         \"VendorStreetAddress\": \"$Vendor:StreetAddress\",\n         \"VendorPOBox\": \"$Vendor:POBox\",\n         \"VendorLocality\": \"$Vendor:Locality\",\n         \"VendorRegion\": \"$Vendor:Region\",\n         \"VendorPostalCode\": \"$Vendor:PostalCode\",\n         \"VendorCountry\": \"$Vendor:Country\"\n       },\n       \"VendorTelephone\": \"$Vendor:Telephone\",\n       \"VendorCustomerID\": \"$Vendor:CustomerID\",\n       \"VendorStatus\": \"$Vendor:Status\",\n       \"VendorSWIFT\": \"$Vendor:SWIFT\",\n       \"VendorIBAN\": \"$Vendor:IBAN\",\n       \"VendorCustom1\": \"$Vendor:Custom1\",\n       \"VendorCustom2\": \"$Vendor:Custom2\",\n       \"VendorCustom3\": \"$Vendor:Custom3\",\n       \"VendorCustom4\": \"$Vendor:Custom4\",\n       \"VendorCustom5\": \"$Vendor:Custom5\",\n       \"VendorMatched\": \"$Vendor:Matched\"\n     }\n}"
```
Comment (bennettgo) [inline custom_connectors/oauth2/ephesoft_connector.rb:92]: One way to make this easier to maintain. I believe you can declare the payload in its natural json format and then stringify it.

```
payload = {
"AmountDue": "$AmountDue"
# more keys here
}

post("https://api.us.ephesoft.io/v1/settings/integrations/configurations",
          integrationName: "Workato Recipe #{recipe_id}",
          integrationType: "webhook",
          enabled: true,
          settings: {
            targetUrl: webhook_url,
            encoding: "application/json",
            payload: payload.to_json.to_s
           }
        )
```

Comment (bennettgo) [reply custom_connectors/oauth2/ephesoft_connector.rb:92]: Might be easier to add more keys if you need to in the future.

File: custom_connectors/oauth2/ephesoft_connector.rb [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+      dedup: lambda do |messages|
+        Time.now.to_f
+      end
+    }
+  },
+  object_definitions: {
+    webhook_output: {
+      fields: lambda do
+        [
+          { name: "AmountDue"},
```
Comment (bennettgo) [inline custom_connectors/oauth2/ephesoft_connector.rb:169]: Output definitions are missing the `type` declarations. 

https://docs.workato.com/developing-connectors/sdk/object-definition.html#structuring-input-and-output-fields

**Review by bennettgo — CHANGES_REQUESTED** [review]
@Ephesoft-Stitus Hey Shaun, few changes requested!

File: custom_connectors/oauth2/ephesoft_connector.rb (hunk 1 above)
Comment (Ephesoft-Stitus) [reply custom_connectors/oauth2/ephesoft_connector.rb]: Removed in latest commit.

File: custom_connectors/oauth2/ephesoft_connector.rb (hunk 2 above)
Comment (Ephesoft-Stitus) [reply custom_connectors/oauth2/ephesoft_connector.rb]: Updated title to follow standard in latest commit.

File: custom_connectors/oauth2/ephesoft_connector.rb (hunk 4 above)
Comment (Ephesoft-Stitus) [reply custom_connectors/oauth2/ephesoft_connector.rb]: Yeah, this is much better.  Updated in latest commit.

File: custom_connectors/oauth2/ephesoft_connector.rb (hunk 5 above)
Comment (Ephesoft-Stitus) [reply custom_connectors/oauth2/ephesoft_connector.rb]: Added types in latest commit.  As an FYI: I was previously following the sample code on the referenced page which only adds the type declarations to non-string types.  All of these are strings which is why I didn't add any earlier.

File: custom_connectors/oauth2/ephesoft_connector.rb (hunk 3 above)
Comment (Ephesoft-Stitus) [reply custom_connectors/oauth2/ephesoft_connector.rb]: Updated title to follow standard convention in latest commit.

**Comment by Ephesoft-Stitus** [comment]
@bennettgo  - requested changes implemented.  Thanks for the suggestions!

**Review by bennettgo — APPROVED** [review]

**Pull request merged**